init.lua

```diff
@@ -89,19 +89,51 @@ function torch.type(obj)
 end
 
 --[[ See if a given object is an instance of the provided torch class. ]]
+local function exactTypeMatch(obj_mt, typeSpecNameString)
+  return obj_mt.__typename == typeSpecNameString
+end
+
+local function partialTypeMatch(obj_mt, typeSpecNameString)
```

Comment:
1 - "effectively just checking that obj_mt.__typename is a suffix of typeSpecNameString ?"
the original bug reported complained that "TorchDummy" is erranously matching "OtherTorchDummy", which exactly the behaviour we are trying to avoid.

typeSpecNameString needs to be the last proper "as in after . separation" type name of obj_mt.__typename. so for example we want obj of type "Top.Middle.Low" to match with typeSpecNameStrings: "Middle.Low" and "Low".

2 - As for the performance issue, I am not sure if the cost of executing string.reverse would be lower than traversing in place, which is free, the tables: objNamesArray and typeSpecNamesArray are needed anyways to get the behaviour outlined in point 1 above.

I am kind of new to Lua in general, so probably this code needs work still.